### ISSUE TITLE
Syncing 'memberof'

### ISSUE THREAD

**Issue by NetworkSeb** [Adldap2/Adldap2-Laravel#29]
Hi All,

I'm trying to sync the memberof user attribute but in my user model it only returns one group (the first one). Is this intentional?

User model:
```
 #original: array:15 [▼
    "id" => "1"
    "username" => "sebtest"
    "student_id" => null
    "staff_id" => "TEST"
    "full_name" => "Seb Test"
    "firstname" => "Seb"
    "email" => "sebtest@example.co.uk"
    "primary_group" => null
    "groups" => "CN=Seb,OU=Email groups,OU=Groups,DC=example,DC=local"
    "manager" => null
    "password" => ""
    "old_password" => null
    "remember_token" => ""
    "created_at" => "2015-11-18 11:55:17"
    "updated_at" => "2015-11-18 11:55:17"
]
```
From AD:
```
"memberof" => array:13 [▼
        0 => "CN=Seb,OU=Email groups,OU=Groups,DC=example,DC=local"
        1 => "CN=Test,OU=Email groups,OU=Groups,DC=example,DC=local"
        2 => "CN=Services,OU=Additional,OU=StudentGroups,OU=Moodle,OU=Groups,OU=example,DC=example,DC=local"
        3 => "CN=Languages,OU=example,OU=Personnel,OU=Groups,DC=example,DC=local"
        4 => "CN=Dept 1,OU=example,OU=Personnel,OU=Groups,DC=example,DC=local"
        5 => "CN=Dept 2,OU=example,OU=Personnel,OU=Groups,DC=example,DC=local"
        6 => "CN=Dept 3,OU=example,OU=Personnel,OU=Groups,DC=example,DC=local"
        7 => "CN=Staff,OU=example,OU=Email groups,OU=Groups,DC=example,DC=local"
        8 => "CN=4,OU=Email groups,OU=Groups,DC=example,DC=local"
        9 => "CN=5,OU=Email groups,OU=Groups,DC=example,DC=local"
        10 => "CN=C,OU=Security,OU=Groups,OU=example,DC=example,DC=local"
        11 => "CN=Managers,OU=Email groups,OU=Groups,DC=example,DC=local"
        12 => "CN=Domain Users,CN=Users,DC=example,DC=local"
      ]
```
Also, is there some documentation on all of the methods I can call on an Adldap object?

Thanks,

NS

**Comment by stevebauman** [comment]
This is intended because we don't know exactly what the Adldap model attribute we're trying to sync contains, so if it's an array it retrieves the first element.

https://github.com/Adldap2/Adldap2-Laravel/blob/master/src/AdldapAuthUserProvider.php#L152

Maybe offering a callback class to retrieve the Adldap attribute might be of use?

```php
'sync_attributes' => [
    'name' => 'App\Handlers\LdapAttributeHandler@name',
]
```

```php
use Adldap\Models\User;

class LdapAttributeHandler
{
    public function name(User $user)
    {
        return $user->getCn();
    }
}
```

Also there's a doc for each model located here:

https://github.com/Adldap2/Adldap2/tree/v5.2/docs/models

You can also look at the API as well:

http://adldap2.github.io/api/v5.2.0/Adldap/Models.html

**Comment by michaelachrisco** [comment]
Having a facade/model attribute casting/etc... of MemberOf would be the last piece to remove the custom LDAP controller in my project for this package. Personally would like Adldap2-Laravel to take care of the membership of users. 

May take a stab at this when I have time :8ball:

**Comment by NetworkSeb** [comment]
I'd love this too!

**Comment by michaelachrisco** [comment]
Just an FYI, if just reading the DB, you can use:
```php
class LDAPAuthController {
     public function userGroups($username)
     {
       if(empty($username)){
         Log::error('Error binding to LDAP: username empty');
         return false;
       }

      $ldapconn=ldap_connect("IP or servername") or die("Could not connect to LDAP server.");
      $result = false;

       if ($ldapconn)
       {
        try{

           // bind the connection
           $bind = @ldap_bind($ldapconn, 'admin','adminpass');

           if($bind){
             $result = ldap_search($ldapconn,"OU=ORG,DC=orgname,DC=local", "mail={$username}", array('distinguishedname', 'memberof'));
             $entries = ldap_get_entries($ldapconn, $result);

             if (empty($entries) || $entries['count']==0){
               ldap_unbind($ldapconn);
               //Incorrect email address
               return [];

             } else{
               // valid credentials
               $group = [];
               $memberOfGroups = $entries[0]['memberof'];
               $LDAPUserGroups = array_except($memberOfGroups, ['count']);
               foreach($LDAPUserGroups as $LDAPGroup){
                 $unformattedUserGroups = explode(',',$LDAPGroup);
                 $group[] = str_replace('CN=', '', $unformattedUserGroups[0]);
               }

                ldap_unbind($ldapconn);
                return $group;
              }
           }
           else{
              // invalid credentials
              ldap_unbind($ldapconn);
              return [];
           }

        }
        //Blow up and Log
        catch(Exception $e){
           Log::error($e);
           return [];
        }


          // ldap_unbind($ldapconn);
       }
       else {
         Log::error('Error connecting to LDAP.');
         return false;
       }
       ldap_unbind($ldapconn);

       }
}
```

This is a (somewhat) direct translation of a Ruby controller so feel free to tweak it.

**Comment by stevebauman** [comment]
Added this in just now:

https://github.com/Adldap2/Adldap2-Laravel/commit/80cbeec41b27b49df41267b767547c3c79f44b6b

**Comment by stevebauman** [comment]
Released `v1.3.8`: https://github.com/Adldap2/Adldap2-Laravel/releases/tag/v1.3.8

**Comment by michaelachrisco** [comment]
Awesome thanks! Ill give it a shot.
